File: src/token.rs
// Copyright (c) 2025, TheByteSlayer, Lithium Cache

use rand::Rng;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::fs;
use std::io;
use std::path::Path;

#[derive(Debug, Serialize, Deserialize)]
pub struct TokenConfig {
    pub token: String,
}

impl TokenConfig {
    pub fn is_valid_token(token: &str) -> bool {
        let pattern = Regex::new(r"^[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}$").unwrap();
        pattern.is_match(token)
    }
    
    pub fn generate_token() -> String {
        let mut rng = rand::thread_rng();
        let mut token_parts = Vec::new();
        
        for _ in 0..4 {
            let part: u16 = rng.r#gen();
            token_parts.push(format!("{:04x}", part));
        }
        
        token_parts.join("-")
    }
    
    pub fn new() -> Self {
        TokenConfig {
            token: Self::generate_token(),
        }
    }
    
    pub fn load_or_create() -> io::Result<Self> {
        let token_path = "token.json";
        
        if Path::new(token_path).exists() {
            let content = fs::read_to_string(token_path)?;
            match serde_json::from_str::<TokenConfig>(&content) {
                Ok(token_config) => {
                    if Self::is_valid_token(&token_config.token) {
                        Ok(token_config)
                    } else {
                        let new_token_config = Self::new();
                        new_token_config.save(token_path)?;
                        Ok(new_token_config)
                    }
                }
                Err(_) => {
                    let new_token_config = Self::new();
                    new_token_config.save(token_path)?;
                    Ok(new_token_config)
                }
            }
        } else {
            let new_token_config = Self::new();
            new_token_config.save(token_path)?;
            Ok(new_token_config)
        }
    }
    
    pub fn save(&self, path: &str) -> io::Result<()> {
        let json_content = serde_json::to_string_pretty(self)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
        fs::write(path, json_content)
    }
}

impl Default for TokenConfig {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use regex::Regex;
    
    #[test]
    fn test_token_format() {
        let token = TokenConfig::generate_token();
        let pattern = Regex::new(r"^[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}$").unwrap();
        assert!(pattern.is_match(&token), "Token '{}' doesn't match expected format", token);
    }
    
    #[test]
    fn test_token_uniqueness() {
        let token1 = TokenConfig::generate_token();
        let token2 = TokenConfig::generate_token();
        assert_ne!(token1, token2, "Generated tokens should be unique");
    }
    
    #[test]
    fn test_token_validation() {
        assert!(TokenConfig::is_valid_token("1a2b-3c4d-5e6f-7f80"));
        assert!(TokenConfig::is_valid_token("0000-0000-0000-0000"));
        assert!(TokenConfig::is_valid_token("ffff-ffff-ffff-ffff"));
        
        assert!(!TokenConfig::is_valid_token("1a2b-3c4d-5e6f"));
        assert!(!TokenConfig::is_valid_token("1a2b-3c4d-5e6f-7f80-1234"));
        assert!(!TokenConfig::is_valid_token("1A2B-3C4D-5E6F-7F80"));
        assert!(!TokenConfig::is_valid_token("1a2b_3c4d_5e6f_7f80"));
        assert!(!TokenConfig::is_valid_token("1a2g-3c4d-5e6f-7f80"));
        assert!(!TokenConfig::is_valid_token(""));
        assert!(!TokenConfig::is_valid_token("not-a-token"));
    }
}

File: src/token.cpp
// Copyright (c) 2025, TheByteSlayer, Lithium Cache

#include "token.hpp"
#include <fstream>
#include <iostream>
#include <regex>
#include <random>
#include <iomanip>
#include <sstream>
#include <filesystem>

namespace lithium {

TokenConfig::TokenConfig() : token(generate_token()) {}

TokenConfig::TokenConfig(const std::string& token) : token(token) {}

bool TokenConfig::is_valid_token(const std::string& token) {
    std::regex pattern(R"(^[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}$)");
    return std::regex_match(token, pattern);
}

std::string TokenConfig::generate_token() {
    std::string token;
    token.reserve(19);
    
    for (int i = 0; i < 4; ++i) {
        if (i > 0) {
            token += "-";
        }
        token += generate_hex_segment();
    }
    
    return token;
}

std::string TokenConfig::generate_hex_segment() {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<std::uint16_t> dis(0, 0xFFFF);
    
    std::uint16_t value = dis(gen);
    std::stringstream ss;
    ss << std::hex << std::setfill('0') << std::setw(4) << value;
    return ss.str();
}

std::optional<TokenConfig> TokenConfig::load_or_create(const std::string& token_path) {
    if (std::filesystem::exists(token_path)) {
        std::ifstream file(token_path);
        if (!file.is_open()) {
            std::cerr << "Failed to open token file: " << token_path << std::endl;
            return std::nullopt;
        }
        
        std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
        file.close();
        
        try {
            nlohmann::json json_token = nlohmann::json::parse(content);
            TokenConfig token_config = json_token.get<TokenConfig>();
            
            if (is_valid_token(token_config.token)) {
                return token_config;
            } else {
                TokenConfig new_token_config;
                if (new_token_config.save(token_path)) {
                    return new_token_config;
                }
                return std::nullopt;
            }
        } catch (const nlohmann::json::exception&) {
            TokenConfig new_token_config;
            if (new_token_config.save(token_path)) {
                return new_token_config;
            }
            return std::nullopt;
        }
    } else {
        TokenConfig new_token_config;
        if (new_token_config.save(token_path)) {
            return new_token_config;
        }
        return std::nullopt;
    }
}

bool TokenConfig::save(const std::string& token_path) const {
    try {
        nlohmann::json json_token = *this;
        std::ofstream file(token_path);
        if (!file.is_open()) {
            std::cerr << "Failed to open token file for writing: " << token_path << std::endl;
            return false;
        }
        
        file << json_token.dump(4) << std::endl;
        file.close();
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Error saving token: " << e.what() << std::endl;
        return false;
    }
}

}
